src/app/tabs/camera.cpp

```diff
@@ -1,135 +1,361 @@
+        default_device = QCameraInfo::defaultCamera().deviceName();
+
+    int i = 0;
+    for (auto const& cam : cameras) {
+      QString pretty_name = cam.description() + " at " + cam.deviceName();
+      local_cams.append(QPair<QString,QString>(pretty_name, cam.deviceName()));
+      if (cam.deviceName() == default_device)
+      this->local_index = i;
+      i++;
+    }
+
+    if (this->local_cams.isEmpty()) {
+        this->local_cams.append(QPair<QString,QString>(QString("<No local cameras found>"),QString()));
+        return false;
+    }
+    return true;
+}
+
+QWidget *CameraTab::input_widget()
+{
+    QWidget *widget = new QWidget(this);
+    QHBoxLayout *layout = new QHBoxLayout(widget);
+
+    QLineEdit *input = new QLineEdit(this->config->get_cam_network_url(), widget);
+    input->setContextMenuPolicy(Qt::NoContextMenu);
+    input->setFont(Theme::font_18);
+    input->setAlignment(Qt::AlignCenter);
+    connect(input, &QLineEdit::textEdited, [this](QString text) { this->config->set_cam_network_url(text); });
+    connect(input, &QLineEdit::returnPressed, this, &CameraTab::connect_network_stream);
+
+    layout->addStretch(1);
+    layout->addWidget(input, 4);
+    layout->addStretch(1);
+
+    return widget;
+}
+
+void CameraTab::update_network_status(QMediaPlayer::MediaStatus media_status)
+{
+    qInfo() << "camera status changed to: " << media_status;
+
+    switch (media_status) {
+        case QMediaPlayer::LoadingMedia:
+        case QMediaPlayer::LoadedMedia:
+        case QMediaPlayer::BufferedMedia:
+            this->status->setText("connecting...");
+            break;
+        default:
+            this->status->setText("connection failed");
+            emit disconnected();
+            break;
+    }
+}
+
+void CameraTab::connect_network_stream()
+{
+    connect(this->player, &QMediaPlayer::mediaStatusChanged,
+            [this](QMediaPlayer::MediaStatus media_status) { this->update_network_status(media_status); });
+    connect(this->player, QOverload<>::of(&QMediaPlayer::metaDataChanged), [this]() { emit connected_network(); });
+    qInfo() << "playing stream: " << this->config->get_cam_network_url();
+    this->player->setMedia(QUrl(this->config->get_cam_network_url()));
+    this->player->play();
+}
+
+void CameraTab::connect_local_stream()
+{
+    if (this->local_cam != nullptr) {
+        delete this->local_cam;
+        this->local_cam = nullptr;
+    }
+
+    const QString& local = this->config->get_cam_local_device();
+    if (!this->local_cam_available(local)) {
+        this->status->setText("Camera unavailable");
+        return;
+    }
+
+    qDebug() << "Connecting to local cam " << local;
+    this->local_cam = new QCamera(local.toUtf8(), this);
+    this->local_cam->setViewfinder(this->local_video_widget);
+    connect(this->local_cam, &QCamera::statusChanged, this, &CameraTab::update_local_status);
+    this->local_cam->start();
+}
+
+void CameraTab::choose_video_resolution()
```

Comment:
so for the local cam you can only use supported resolutions? I'm assuming any additional space will be black space?

Reply:
It uses the highest resolution that will fit on the screen

Reply:
The algorithm picks the resolution with the least amount of blank space (sort of).

Reply:
The camera only outputs specific resolutions. You can stretch them to fit but I am prioritizing dimensional accuracy over filling the screen.